ocaml/xapi/repository.ml

```diff
@@ -188,58 +231,136 @@ let sync ~__context ~self ~token ~token_id =
         in
-        let Pkg_mgr.{cmd; params} =
-          [
-            "--save"
-          ; proxy_url_param
-          ; proxy_username_param
-          ; proxy_password_param
-          ; token_param
-          ]
-          |> fun config -> Pkgs.config_repo ~repo_name ~config
+
+        (* Sync with remote repository *)
+        let sync_repo () =
+          let Pkg_mgr.{cmd; params} = Pkgs.sync_repo ~repo_name in
+          Unixext.mkdir_rec !Xapi_globs.local_pool_repo_dir 0o700 ;
+          clean_yum_cache repo_name ;
+          ignore (Helpers.call_script cmd params)
+        in
+
+        let server_auth conf f =
+          match conf with
+          | DefaultAuth | NoAuth ->
+              f ()
+          | ServerAuthConf {cert; remote_addr; remote_port} ->
+              let ( let@ ) f x = f x in
+              let@ temp_file =
+                Helpers.with_temp_file_of_content "external-host-cert-" ".pem"
+                  cert
+              in
+              Stunnel.with_client_proxy
+                ~verify_cert:(Stunnel_client.external_host temp_file)
+                ~remote_host:remote_addr ~remote_port ~local_host ~local_port
+              @@ fun () -> f ()
         in
-        ignore (Helpers.call_script ~log_output:Helpers.On_failure cmd params) ;
 
-        (* Import YUM repository GPG key to check metadata in reposync *)
-        let Pkg_mgr.{cmd; params} = Pkgs.make_cache ~repo_name in
-        ignore (Helpers.call_script cmd params) ;
+        let auth =
+          match client_auth_config with
+          | CdnTokenAuthConf {token; token_id} ->
+              Some (CdnTokenAuth (token_id, token), "accesstoken")
+          | PoolExtHostAuthConf {cert; remote_addr; username; password} ->
+              let verified_rpc =
+                try
+                  Helpers.make_external_host_verified_rpc ~__context remote_addr
+                    cert
+                with Xmlrpc_client.Connection_reset ->
+                  raise
+                    (Api_errors.Server_error
+                       ( Api_errors
+                         .update_syncing_remote_pool_coordinator_connection_failed
+                       , []
+                       )
+                    )
+              in
+              let session_id =
+                try
+                  Client.Client.Session.login_with_password ~rpc:verified_rpc
+                    ~uname:username ~pwd:password
+                    ~version:Datamodel_common.api_version_string
+                    ~originator:Xapi_version.xapi_user_agent
+                with
+                | Http_client.Http_request_rejected _ | Http_client.Http_error _
+                ->
+                  raise
+                    (Api_errors.Server_error
+                       ( Api_errors
+                         .update_syncing_remote_pool_coordinator_service_failed
+                       , []
+                       )
+                    )
+              in
```

Comment:
This part can be generalized as a function being like `Helpers.login_after_server_check`.